<a href="https://colab.research.google.com/github/Isaac177/cocktail/blob/main/final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
!rm -rf /content/drive
!mkdir /content/drive

In [20]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os
print("Files in Google Drive root:")
print(os.listdir('/content/drive/MyDrive/Colab Notebooks'))

Files in Google Drive root:
['Copy of COLAB - CIFAR10_image_classification.ipynb', 'Neural-1.ipynb', 'Assignement 3.ipynb', 'Assignment_2.ipynb', 'Copy of cifar10_tutorial.ipynb', 'archive.zip', 'Untitled0.ipynb', 'Untitled2.ipynb', 'extracted_archive', 'Untitled1.ipynb', 'final.ipynb']


In [ ]:
import zipfile

# Updated path to your zip file
zip_path = '/content/drive/MyDrive/Colab Notebooks/archive.zip'

# Create a directory to extract to
extract_path = '/content/drive/MyDrive/Colab Notebooks/extracted_archive'

# Create the directory if it doesn't exist
!mkdir -p "{extract_path}"

# Extract the contents
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_path)

# Let's see what was extracted
print("\nExtracted contents:")
print(os.listdir(extract_path))

In [4]:
import os

# Let's look inside the dogcat folder
dogcat_path = '/content/drive/MyDrive/Colab Notebooks/extracted_archive/dogcat'
print("Contents of dogcat folder:")
print(os.listdir(dogcat_path))

# Let's also count the number of files and check a few sample paths
sample_files = os.listdir(dogcat_path)[:5]  # Show first 5 files
print("\nFirst 5 files:")
for file in sample_files:
    print(file)

# Count total files
total_files = len(os.listdir(dogcat_path))
print(f"\nTotal number of files: {total_files}")

Contents of dogcat folder:
['test1', 'train', 'validation', 'sampleSubmission.csv']

First 5 files:
test1
train
validation
sampleSubmission.csv

Total number of files: 4


In [5]:
# Check contents of each directory
base_path = '/content/drive/MyDrive/Colab Notebooks/extracted_archive/dogcat'

# Check train folder
train_path = os.path.join(base_path, 'train')
print("Number of files in train folder:", len(os.listdir(train_path)))
print("Sample train files:", os.listdir(train_path)[:5])

# Check validation folder
val_path = os.path.join(base_path, 'validation')
print("\nNumber of files in validation folder:", len(os.listdir(val_path)))
print("Sample validation files:", os.listdir(val_path)[:5])

# Check test folder
test_path = os.path.join(base_path, 'test1')
print("\nNumber of files in test folder:", len(os.listdir(test_path)))
print("Sample test files:", os.listdir(test_path)[:5])


Number of files in train folder: 2
Sample train files: ['cats', 'dogs']

Number of files in validation folder: 2
Sample validation files: ['cats', 'dogs']

Number of files in test folder: 1
Sample test files: ['test1']


In [6]:
import os

# Check the number of images in each folder
base_path = '/content/drive/MyDrive/Colab Notebooks/extracted_archive/dogcat'

# Count images in train folders
train_cats = len(os.listdir(os.path.join(base_path, 'train/cats')))
train_dogs = len(os.listdir(os.path.join(base_path, 'train/dogs')))

# Count images in validation folders
val_cats = len(os.listdir(os.path.join(base_path, 'validation/cats')))
val_dogs = len(os.listdir(os.path.join(base_path, 'validation/dogs')))

# Count test images
test_images = len(os.listdir(os.path.join(base_path, 'test1/test1')))

print("Training Data:")
print(f"Cats: {train_cats}, Dogs: {train_dogs}")
print("\nValidation Data:")
print(f"Cats: {val_cats}, Dogs: {val_dogs}")
print("\nTest Data:")
print(f"Total test images: {test_images}")

Training Data:
Cats: 12500, Dogs: 12500

Validation Data:
Cats: 4000, Dogs: 4000

Test Data:
Total test images: 12500


In [7]:
# Import necessary libraries
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.applications import VGG16
from tensorflow.keras import layers, models
import matplotlib.pyplot as plt
import numpy as np
import os

# Set random seed for reproducibility
tf.random.set_seed(42)

# Define basic constants
IMG_HEIGHT = 224
IMG_WIDTH = 224
BATCH_SIZE = 32
BASE_PATH = '/content/drive/MyDrive/Colab Notebooks/extracted_archive/dogcat'

# Verify paths and print dataset information
print("Checking dataset structure...")
print("\nTraining Data:")
print(f"Cats: {len(os.listdir(os.path.join(BASE_PATH, 'train/cats')))}")
print(f"Dogs: {len(os.listdir(os.path.join(BASE_PATH, 'train/dogs')))}")
print("\nValidation Data:")
print(f"Cats: {len(os.listdir(os.path.join(BASE_PATH, 'validation/cats')))}")
print(f"Dogs: {len(os.listdir(os.path.join(BASE_PATH, 'validation/dogs')))}")

Checking dataset structure...

Training Data:
Cats: 12500
Dogs: 12500

Validation Data:
Cats: 4000
Dogs: 4000


In [8]:
# Set up data generators with augmentation for training data
train_datagen = ImageDataGenerator(
    rescale=1./255,  # Normalize pixel values
    rotation_range=20,  # Randomly rotate images by up to 20 degrees
    width_shift_range=0.2,  # Randomly shift images horizontally
    height_shift_range=0.2,  # Randomly shift images vertically
    shear_range=0.2,  # Shear transformations
    zoom_range=0.2,  # Random zoom
    horizontal_flip=True,  # Randomly flip images horizontally
    fill_mode='nearest'  # Strategy for filling in newly created pixels
)

# Only rescaling for validation data (no augmentation needed)
validation_datagen = ImageDataGenerator(rescale=1./255)

# Create generators for training and validation
train_generator = train_datagen.flow_from_directory(
    os.path.join(BASE_PATH, 'train'),
    target_size=(IMG_HEIGHT, IMG_WIDTH),
    batch_size=BATCH_SIZE,
    class_mode='binary'
)

validation_generator = validation_datagen.flow_from_directory(
    os.path.join(BASE_PATH, 'validation'),
    target_size=(IMG_HEIGHT, IMG_WIDTH),
    batch_size=BATCH_SIZE,
    class_mode='binary'
)

# Let's look at the generator settings
print("\nTraining Generator:")
print(f"Found {train_generator.samples} images belonging to {len(train_generator.class_indices)} classes")
print("\nValidation Generator:")
print(f"Found {validation_generator.samples} images belonging to {len(validation_generator.class_indices)} classes")

Found 25000 images belonging to 2 classes.
Found 8000 images belonging to 2 classes.

Training Generator:
Found 25000 images belonging to 2 classes

Validation Generator:
Found 8000 images belonging to 2 classes


In [9]:
# Load the pre-trained VGG16 model
base_model = VGG16(
    weights='imagenet',  # Use pre-trained weights
    include_top=False,   # Don't include the final classification layers
    input_shape=(IMG_HEIGHT, IMG_WIDTH, 3)  # Specify input image size
)

# Freeze the pre-trained layers
for layer in base_model.layers:
    layer.trainable = False

# Create our model
model = models.Sequential([
    # Base VGG16 model
    base_model,

    # Add our classification layers
    layers.Flatten(),
    layers.Dense(512, activation='relu'),
    layers.Dropout(0.5),
    layers.Dense(1, activation='sigmoid')  # Binary classification (dog or cat)
])

# Display model summary
model.summary()

58889256/58889256 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ vgg16 (Functional)                   │ (None, 7, 7, 512)           │      14,714,688 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 25088)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 512)                 │      12,845,568 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 512)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 1)                   │             513 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 27,560,769 (105.14 MB)

 Trainable params: 12,846,081 (49.00 MB)

 Non-trainable params: 14,714,688 (56.13 MB)

In [ ]:
#Step 3: Model Compilation and Training


model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

# Add early stopping to prevent overfitting
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor='val_loss',
    patience=3,
    restore_best_weights=True
)

# Train the model
history = model.fit(
    train_generator,
    epochs=2,
    validation_data=validation_generator,
    callbacks=[early_stopping]
)

Epoch 1/2
163/782 ━━━━━━━━━━━━━━━━━━━━ 2:23:37 14s/step - accuracy: 0.7405 - loss: 0.7099

In [9]:
# Function to plot training history
def plot_training_results(history):
    # Create figure with two subplots
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))

    # Plot accuracy
    ax1.plot(history.history['accuracy'], label='Training Accuracy')
    ax1.plot(history.history['val_accuracy'], label='Validation Accuracy')
    ax1.set_title('Model Accuracy')
    ax1.set_xlabel('Epoch')
    ax1.set_ylabel('Accuracy')
    ax1.legend()

    # Plot loss
    ax2.plot(history.history['loss'], label='Training Loss')
    ax2.plot(history.history['val_loss'], label='Validation Loss')
    ax2.set_title('Model Loss')
    ax2.set_xlabel('Epoch')
    ax2.set_ylabel('Loss')
    ax2.legend()

    plt.tight_layout()
    plt.show()

# Plot the results (run this after training completes)
plot_training_results(history)

# Print final metrics
final_accuracy = history.history['val_accuracy'][-1]
final_loss = history.history['val_loss'][-1]
print(f"\nFinal Validation Accuracy: {final_accuracy:.4f}")
print(f"Final Validation Loss: {final_loss:.4f}")

NameError: name 'history' is not defined

In [ ]:
# Function to load and preprocess a single image
def preprocess_image(img_path):
    img = tf.keras.preprocessing.image.load_img(
        img_path,
        target_size=(IMG_HEIGHT, IMG_WIDTH)
    )
    img_array = tf.keras.preprocessing.image.img_to_array(img)
    img_array = img_array / 255.0  # Normalize
    img_array = tf.expand_dims(img_array, 0)
    return img_array

# Function to make prediction
def predict_image(img_path):
    # Preprocess image
    img_array = preprocess_image(img_path)

    # Make prediction
    prediction = model.predict(img_array)

    # Convert prediction to class label
    class_label = 'Dog' if prediction[0] > 0.5 else 'Cat'
    confidence = prediction[0] if prediction[0] > 0.5 else 1 - prediction[0]

    return class_label, confidence

# Let's test on a few images from the test set
test_folder = os.path.join(BASE_PATH, 'test1/test1')
test_images = os.listdir(test_folder)[:5]  # Get first 5 test images

for image_name in test_images:
    image_path = os.path.join(test_folder, image_name)

    # Make prediction
    predicted_class, confidence = predict_image(image_path)

    # Display image and prediction
    img = plt.imread(image_path)
    plt.figure(figsize=(6, 6))
    plt.imshow(img)
    plt.title(f'Prediction: {predicted_class} (Confidence: {confidence:.2%})')
    plt.axis('off')
    plt.show()

In [ ]:
# Save the model
model_save_path = '/content/drive/MyDrive/Colab Notebooks/dogcat_model'
model.save(model_save_path)
print(f"Model saved to {model_save_path}")

# Save the training history as a numpy file
import numpy as np
np.save('/content/drive/MyDrive/Colab Notebooks/training_history.npy', history.history)
print("Training history saved")

# Computer Vision Final Project: Dogs vs Cats Classification
## By [Your Name]

### 1. Problem Understanding and Dataset Selection (10%)
#### Problem Definition
This project addresses the fundamental computer vision task of binary image classification, specifically distinguishing between images of dogs and cats. This problem is significant because:
- It demonstrates the capability of deep learning in recognizing complex visual patterns
- It serves as a benchmark for testing various CNN architectures
- It has practical applications in pet-related applications and animal monitoring systems

#### Dataset Description
- **Source**: Dogs vs Cats dataset from Kaggle
- **Size and Distribution**:
  - Total Images: 45,000
  - Training Set: 25,000 (12,500 dogs, 12,500 cats)
  - Validation Set: 8,000 (4,000 dogs, 4,000 cats)
  - Test Set: 12,500 images
- **Image Properties**: Color images of varying sizes and aspects ratios
- **Dataset Quality**: High-quality images with clear subjects and diverse backgrounds

### 2. Data Preprocessing (20%)
#### Image Preprocessing
1. **Resizing**:
   - Standardized all images to 224x224 pixels
   - Chosen to match VGG16's input requirements
   - Preserves important features while managing computational resources

2. **Normalization**:
   - Scaled pixel values to range [0,1]
   - Improves training stability and convergence

#### Data Augmentation
Implemented following transformations to prevent overfitting:
- Rotation (±20 degrees)
- Width/height shifts (20%)
- Shear transformation (20%)
- Zoom range (20%)
- Horizontal flips
- Fill mode: 'nearest'

#### Data Organization
- Structured directory hierarchy for train/validation/test sets
- Implemented data generators for efficient memory usage
- Maintained balanced class distribution

### 3. Model Selection and Implementation (25%)
#### Architecture Choice
Selected VGG16 as the base model for transfer learning because:
- Proven architecture for image classification
- Pre-trained on ImageNet dataset
- Good balance between complexity and performance

#### Model Architecture
1. **Base Model**:
   - VGG16 (excluding top layers)
   - Frozen weights to preserve learned features

2. **Custom Layers**:
   - Flatten layer to convert features to 1D
   - Dense layer (512 units, ReLU activation)
   - Dropout (0.5) for regularization
   - Output layer (1 unit, sigmoid activation)

3. **Parameters**:
   - Total Parameters: 27,560,769
   - Trainable Parameters: 12,846,081
   - Non-trainable Parameters: 14,714,688

### 4. Model Training and Evaluation (30%)
#### Training Configuration
- **Optimizer**: Adam
- **Loss Function**: Binary Cross-entropy
- **Metrics**: Accuracy
- **Batch Size**: 32
- **Epochs**: 15 with early stopping
- **Early Stopping**: Patience=3, monitoring validation loss

#### Training Results
[Add your actual training results here:
- Final training accuracy
- Final validation accuracy
- Training/validation loss curves]

#### Evaluation Metrics
[Add your model's performance metrics here:
- Accuracy
- Sample predictions on test images]

### 5. Result Analysis and Discussion (15%)
#### Model Performance
[Discuss your model's performance, including:
- Analysis of accuracy and loss curves
- Discussion of any overfitting/underfitting
- Analysis of misclassified images]

#### Limitations
1. Limited to binary classification
2. Dependent on image quality
3. Computational requirements of VGG16

#### Future Improvements
1. Experiment with other architectures (ResNet, EfficientNet)
2. Implement cross-validation
3. Try different learning rates and optimizers
4. Add more dense layers
5. Explore ensemble methods

#### Conclusion
[Add your concluding remarks about the project's success and learning outcomes]